### ISSUE TITLE
REPL does not put source locations on `#'`ed forms

### ISSUE THREAD

**Issue by david-christiansen** [greghendershott/racket-mode#556]
It looks like syntax-quoted strings don't get source positions put on them when sent either from the `racket-mode` REPL or by using features like `C-M-x`. They work fine when running source buffers.

The difference can be seen in these interactions from, respectively, racket-mode and DrRacket:
```
racket-mode> (syntax-line #'"foo")
#f

DrRacket> (syntax-line #'"foo")
5
```

This comes up when using `parse-syntax-string` with `syntax/p` to test Megaparsack parsers at the REPL, or when using similar features to test a generated LR parser from `parser-tools`. Megaparsack in particular fails because it attempts to do arithmetic on `#f`. Putting the tests into a test submodule and running it the usual way is a workaround, but a REPL is nice both for developing those tests and for general experimentation and demonstrations.

I suspect that, somewhere, an input port needs to have its location tracking enabled, but I couldn't find the spot in question myself.

Output of `racket-bug-report`:
<details>
<pre>
((alist-get 'racket-mode package-alist)
 #s(package-desc racket-mode
                 (20210831 2045)
                 "Racket editing, REPL, and more"
                 ((emacs
                   (25 1))
                  (faceup
                   (0 0 2))
                  (pos-tip
                   (20191127 1028)))
                 nil nil "/home/dtc/.emacs.d/elpa/racket-mode-20210831.2045"
                 ((:url . "https://www.racket-mode.com/")
                  (:maintainer "Greg Hendershott")
                  (:authors
                   ("Greg Hendershott"))
                  (:commit . "a879a8d67b062c9aa0c6e60602eedba80a2069c5"))
                 nil))
((emacs-version "28.0.50")
 (system-type gnu/linux)
 (x-gtk-use-system-tooltips t)
 (major-mode racket-repl-mode)
 (racket--el-source-dir "/home/dtc/.emacs.d/elpa/racket-mode-20210831.2045/")
 (racket--rkt-source-dir "/home/dtc/.emacs.d/elpa/racket-mode-20210831.2045/racket/")
 (racket-program "racket")
 (racket-command-timeout 10)
 (racket-path-from-emacs-to-racket-function identity)
 (racket-path-from-racket-to-emacs-function identity)
 (racket-browse-url-function racket-browse-url-using-temporary-file)
 (racket-documentation-search-location "https://docs.racket-lang.org/search/index.html?q=%s")
 (racket-xp-after-change-refresh-delay 1)
 (racket-xp-mode-lighter
  (:eval
   (racket--xp-mode-lighter)))
 (racket-xp-highlight-unused-regexp "^[^_]")
 (racket-repl-buffer-name-function nil)
 (racket-submodules-to-run
  ((test)
   (main)))
 (racket-memory-limit 2048)
 (racket-error-context medium)
 (racket-repl-history-directory "~/.emacs.d/racket-mode/")
 (racket-history-filter-regexp "\\`\\s *\\'")
 (racket-images-inline t)
 (racket-imagemagick-props nil)
 (racket-images-keep-last 100)
 (racket-images-system-viewer "display")
 (racket-pretty-print t)
 (racket-use-repl-submit-predicate nil)
 (racket-pretty-print t)
 (racket-indent-curly-as-sequence t)
 (racket-indent-sequence-depth 0)
 (racket-pretty-lambda nil)
 (racket-smart-open-bracket-enable t)
 (racket-module-forms "\\s(\\(?:module[*+]?\\|library\\)")
 (racket-logger-config
  ((cm-accomplice . warning)
   (GC . info)
   (module-prefetch . warning)
   (optimizer . info)
   (racket/contract . error)
   (sequence-specialization . info)
   (* . fatal)))
 (racket-show-functions
  (racket-show-header-line)))
(enabled-minor-modes
 (auto-composition-mode)
 (auto-compression-mode)
 (auto-encryption-mode)
 (auto-fill-mode)
 (auto-save-mode)
 (blink-cursor-mode)
 (column-number-mode)
 (company-mode)
 (file-name-shadow-mode)
 (flycheck-pos-tip-mode)
 (font-lock-mode)
 (global-company-mode)
 (global-eldoc-mode)
 (global-font-lock-mode)
 (global-git-commit-mode)
 (guide-key-mode)
 (line-number-mode)
 (magit-auto-revert-mode)
 (magit-todos-mode)
 (marginalia-mode)
 (mouse-wheel-mode)
 (override-global-mode)
 (paredit-mode)
 (prescient-persist-mode)
 (projectile-mode)
 (recentf-mode)
 (save-place-mode)
 (selectrum-mode)
 (selectrum-prescient-mode)
 (semantic-minor-modes-format)
 (shell-dirtrack-mode)
 (show-paren-mode)
 (tooltip-mode)
 (transient-mark-mode)
 (type-break-mode)
 (type-break-mode-line-message-mode)
 (type-break-query-mode)
 (which-function-mode)
 (windmove-mode))
(disabled-minor-modes
 (abbrev-mode)
 (aggressive-indent-mode)
 (archive-subfile-mode)
 (auto-fill-function)
 (auto-revert-mode)
 (auto-revert-tail-mode)
 (auto-save-visited-mode)
 (buffer-face-mode)
 (buffer-read-only)
 (bug-reference-mode)
 (bug-reference-prog-mode)
 (button-mode)
 (cider--debug-mode)
 (cider-auto-test-mode)
 (cider-enlighten-mode)
 (cider-mode)
 (cider-popup-buffer-mode)
 (cl-old-struct-compat-mode)
 (coffee-cos-mode)
 (coffee-live-compile-mode)
 (company-search-mode)
 (compilation-minor-mode)
 (compilation-shell-minor-mode)
 (completion-in-region-mode)
 (context-menu-mode)
 (dash-fontify-mode)
 (defining-kbd-macro)
 (diff-auto-refine-mode)
 (diff-minor-mode)
 (dired-hide-details-mode)
 (dired-isearch-filenames-mode)
 (display-line-numbers-mode)
 (doc-view-minor-mode)
 (doc-view-presentation-mode)
 (edebug-mode)
 (eglot--managed-mode)
 (eldoc-mode)
 (electric-indent-mode)
 (electric-layout-mode)
 (electric-quote-mode)
 (elisp-slime-nav-mode)
 (ert--current-run-stats)
 (flycheck-mode)
 (flymake-mode)
 (flyspell-mode)
 (git-commit-mode)
 (global-aggressive-indent-mode)
 (global-auto-revert-mode)
 (global-dash-fontify-mode)
 (global-display-line-numbers-mode)
 (global-flycheck-mode)
 (global-hl-todo-mode)
 (global-prettify-symbols-mode)
 (global-semantic-highlight-edits-mode)
 (global-semantic-highlight-func-mode)
 (global-semantic-show-parser-state-mode)
 (global-semantic-show-unmatched-syntax-mode)
 (global-semantic-stickyfunc-mode)
 (global-visual-line-mode)
 (global-whitespace-mode)
 (global-whitespace-newline-mode)
 (gnus-dead-summary-mode)
 (gnus-undo-mode)
 (haskell-collapse-mode)
 (haskell-indentation-mode)
 (highlight-uses-mode)
 (hl-todo-mode)
 (horizontal-scroll-bar-mode)
 (hs-minor-mode)
 (ibuffer-auto-mode)
 (ido-everywhere)
 (idris-simple-indent-mode)
 (image-minor-mode)
 (indent-tabs-mode)
 (interactive-haskell-mode)
 (isearch-mode)
 (ispell-minor-mode)
 (jit-lock-debug-mode)
 (lock-file-mode)
 (lsp-installation-buffer-mode)
 (lsp-managed-mode)
 (lsp-mode)
 (lsp-signature-mode)
 (magit-blame-mode)
 (magit-blame-read-only-mode)
 (magit-blob-mode)
 (magit-wip-after-apply-mode)
 (magit-wip-after-save-local-mode)
 (magit-wip-after-save-mode)
 (magit-wip-before-change-mode)
 (magit-wip-initial-backup-mode)
 (magit-wip-mode)
 (mail-abbrevs-mode)
 (markdown-live-preview-mode)
 (mc-hide-unmatched-lines-mode)
 (meghanada-mode)
 (menu-bar-mode)
 (minibuffer-electric-default-mode)
 (mml-mode)
 (multiple-cursors-mode)
 (next-error-follow-minor-mode)
 (omnisharp-mode)
 (org-cdlatex-mode)
 (org-list-checkbox-radio-mode)
 (org-src-mode)
 (org-table-follow-field-mode)
 (org-table-header-line-mode)
 (orgtbl-mode)
 (outline-minor-mode)
 (outorg-edit-minor-mode)
 (outshine-mode)
 (overwrite-mode)
 (pandoc-mode)
 (paragraph-indent-minor-mode)
 (pcre-mode)
 (popwin-mode)
 (prettify-symbols-mode)
 (racket-smart-open-bracket-mode)
 (racket-xp-mode)
 (rectangle-mark-mode)
 (rectangular-region-mode)
 (rich-minority-mode)
 (rxt--read-pcre-mode)
 (rxt-global-mode)
 (rxt-mode)
 (semantic-highlight-edits-mode)
 (semantic-highlight-func-mode)
 (semantic-mode)
 (semantic-show-parser-state-mode)
 (semantic-show-unmatched-syntax-mode)
 (semantic-stickyfunc-mode)
 (server-mode)
 (sh-electric-here-document-mode)
 (shell-command-with-editor-mode)
 (size-indication-mode)
 (slime-edit-value-mode)
 (slime-editing-mode)
 (slime-macroexpansion-minor-mode)
 (slime-mode)
 (slime-popup-buffer-mode)
 (smerge-mode)
 (tab-bar-history-mode)
 (tab-bar-mode)
 (temp-buffer-resize-mode)
 (text-scale-mode)
 (tool-bar-mode)
 (tracking-mode)
 (transient-resume-mode)
 (unify-8859-on-decoding-mode)
 (unify-8859-on-encoding-mode)
 (url-handler-mode)
 (use-hard-newlines)
 (vc-parent-buffer)
 (view-mode)
 (visible-mode)
 (visual-line-mode)
 (whitespace-mode)
 (whitespace-newline-mode)
 (window-divider-mode)
 (with-editor-mode)
 (ws-butler-global-mode)
 (ws-butler-mode)
 (xref-etags-mode)
 (yas-global-mode)
 (yas-minor-mode))
</pre>
</details>

**Comment by greghendershott** [comment]
Thanks for reporting.

This surprised me because I recall adding ` port-count-lines!` on the REPL's TCP input port for #519.

It looks like this is a result of commit 7f12cb1, which eagerly loads `racket/gui/base` if available (if not on a minimal Racket lacking `gui-lib`). 

What's happening is that `racket/gui/base` sets `current-get-interaction-port` to wrap the original port with a filtering/yielding port. Somehow that's defeating the `port-count-lines!' enabled on the original port. As a result on a system with `gui-lib`, the problem you're reporting happens, and also it undoes the fix for #519. (Running on a minimal Racket, both work correctly.)

TL;DR: I'll take a look and figure out how to make the gui wrapped interaction port do `port-count-lines!`.

---

p.s. As it happens I'm working on #555 right now, which is also related to commit 7f12cb1, but causing a different problem. I think I'm almost ready to merge that, so I'll probably wait to tackle this after rather than juggle both in parallel.

**Comment by david-christiansen** [comment]
Sorry for the slow feedback - I just tested the updated version, and the issue is fixed. Thank you!
